### PR TITLE
Added PVCNN Model 

### PR DESCRIPTION
**Pull request by sanskar107**
Added PVCNN model (both tensorflow and torch)

---
This change is [Reviewable](https://reviewable.io/reviews/isl-org/open3d-ml/394)

### CODE REVIEW

**Review by ssheorey — COMMENTED** [review]
Add expected dtypes and shapes for all inputs and return values in the docstrings. 
Use the new Numpy RNG as in PR #395 

> *[Reviewable](https://reviewable.io/reviews/isl-org/open3d-ml/394#-:-MnMR-wh0grO1wlSXRQe:b-njjt75)* status: 0 of 7 files reviewed, 9 unresolved discussions (waiting on @benjaminum, @sanskar107, and @ssheorey)

___
*[ml3d/tf/models/pvcnn.py, line 17 at r2](https://reviewable.io/reviews/isl-org/open3d-ml/394#-MnMYLm05b0vdSGQOO4d:-MnMYLm05b0vdSGQOO4e:bjp136r) ([raw file](https://github.com/isl-org/open3d-ml/blob/835c152fc729f9aee6159ea2b23984a62a49f0d0/ml3d/tf/models/pvcnn.py#L17)):*
> ```Python
> def trilinear_devoxelize(features, coords, resolution, is_training=True):
>     if not open3d.core.cuda.device_count() > 0:
>         raise NotImplementedError
> ```

Add error message.
___
*[ml3d/tf/models/pvcnn.py, line 106 at r2](https://reviewable.io/reviews/isl-org/open3d-ml/394#-MnMZJRC5FiDfO7ZN0FI:-MnMZJRC5FiDfO7ZN0FJ:bvkfrcs) ([raw file](https://github.com/isl-org/open3d-ml/blob/835c152fc729f9aee6159ea2b23984a62a49f0d0/ml3d/tf/models/pvcnn.py#L106)):*
> ```Python
> 
>         Returns:
>             Returns the probability distribution.
> ```

Add dtype and shape for return value. Is it (B, N, C)?
___
*[ml3d/tf/models/pvcnn.py, line 112 at r2](https://reviewable.io/reviews/isl-org/open3d-ml/394#-MnMZ-6u4tghhG9rlMqj:-MnMZ-6u4tghhG9rlMqk:b-xsui65) ([raw file](https://github.com/isl-org/open3d-ml/blob/835c152fc729f9aee6159ea2b23984a62a49f0d0/ml3d/tf/models/pvcnn.py#L112)):*
> ```
> # coords = inputs[:, :3, :]
> ```
Is xyz dim1 or dim2? Update comment or docstring.
___
*[ml3d/torch/models/pvcnn.py, line 28 at r2](https://reviewable.io/reviews/isl-org/open3d-ml/394#-MnMRBHCBqV9T_VmdwUS:-MnMRBHCBqV9T_VmdwUT:b-b0lphl) ([raw file](https://github.com/isl-org/open3d-ml/blob/835c152fc729f9aee6159ea2b23984a62a49f0d0/ml3d/torch/models/pvcnn.py#L28)):*
> ```Python
>         :param is_training: bool, training mode
>         :return:
>             FloatTensor[B, C, N]
> ```

Update docstrings to Google convention.
___
*[ml3d/torch/models/pvcnn.py, line 137 at r2](https://reviewable.io/reviews/isl-org/open3d-ml/394#-MnMWChhCYdTAaWhZC6m:-MnMWChhCYdTAaWhZC6n:bgpybzn) ([raw file](https://github.com/isl-org/open3d-ml/blob/835c152fc729f9aee6159ea2b23984a62a49f0d0/ml3d/torch/models/pvcnn.py#L137)):*
> ```Python
> 
>         Returns:
>             Returns the probability distribution.
> ```

Mention shape of output tensor. Is it `(B, N, C)`?
___
*[ml3d/torch/models/pvcnn.py, line 143 at r2](https://reviewable.io/reviews/isl-org/open3d-ml/394#-MnMVmtnBhFvC7Q9V0lP:-MnMVmtnBhFvC7Q9V0lQ:b-4h4bb2) ([raw file](https://github.com/isl-org/open3d-ml/blob/835c152fc729f9aee6159ea2b23984a62a49f0d0/ml3d/torch/models/pvcnn.py#L143)):*
> ```
> # coords = inputs[:, :3, :]
> ```
Are XYZ in dim 1 or dim 2? Update comment or docstring.
___
*[ml3d/torch/models/pvcnn.py, line 188 at r2](https://reviewable.io/reviews/isl-org/open3d-ml/394#-MnMWhWO8l5z6lDK0QvK:-MnMWhWO8l5z6lDK0QvL:bebla3f) ([raw file](https://github.com/isl-org/open3d-ml/blob/835c152fc729f9aee6159ea2b23984a62a49f0d0/ml3d/torch/models/pvcnn.py#L188)):*
> ```
> feat = feat / 255.0
> ```
Needs explanation.
___
*[ml3d/torch/models/pvcnn.py, line 201 at r2](https://reviewable.io/reviews/isl-org/open3d-ml/394#-MnMWmQi72jzwajgr5S7:-MnMWmQi72jzwajgr5S8:b-fq9uqj) ([raw file](https://github.com/isl-org/open3d-ml/blob/835c152fc729f9aee6159ea2b23984a62a49f0d0/ml3d/torch/models/pvcnn.py#L201)):*
> ```Python
>         feat = np.concatenate([x, y, z, feat, norm_x, norm_y, norm_z], axis=-1)
> 
>         choices = np.random.choice(
> ```

Update with Numpy RNG PR and use `self.rng`.
___
*[ml3d/torch/models/pvcnn.py, line 273 at r2](https://reviewable.io/reviews/isl-org/open3d-ml/394#-MnMXJmG3JN6CntnF9rc:-MnMXHne0VUOBO2ttQv_:b-3i9om6) ([raw file](https://github.com/isl-org/open3d-ml/blob/835c152fc729f9aee6159ea2b23984a62a49f0d0/ml3d/torch/models/pvcnn.py#L273)):*
> ```Python
> 
>         Attributes:
>             Loss: Object of type `SemSegLoss`.
> ```

Loss -> sem_seg_loss

**Review by sanskar107 — COMMENTED** [review]
@benjaminum This model is not supported on CPU, so the tests won't run on CI. Although, I will add tests for all other unsupported models in a separate PR so that we can run the tests locally.

> *[Reviewable](https://reviewable.io/reviews/isl-org/open3d-ml/394#-:-MnjlZXX6ii7ay5E5h_X:b-3v0yxg)* status: 0 of 34 files reviewed, 9 unresolved discussions (waiting on @benjaminum and @ssheorey)

___
*[ml3d/tf/models/pvcnn.py, line 17 at r2](https://reviewable.io/reviews/isl-org/open3d-ml/394#-MnMYLm05b0vdSGQOO4d:-MnjlB877oFyV-ex9DwQ:b-896fix) ([raw file](https://github.com/isl-org/open3d-ml/blob/835c152fc729f9aee6159ea2b23984a62a49f0d0/ml3d/tf/models/pvcnn.py#L17)):*
<details><summary><i>Previously, ssheorey wrote…</i></summary><blockquote>

Add error message.
</blockquote></details>

Done.
___
*[ml3d/tf/models/pvcnn.py, line 106 at r2](https://reviewable.io/reviews/isl-org/open3d-ml/394#-MnMZJRC5FiDfO7ZN0FI:-MnjlBt0BH7d5qqw51rp:b-896fix) ([raw file](https://github.com/isl-org/open3d-ml/blob/835c152fc729f9aee6159ea2b23984a62a49f0d0/ml3d/tf/models/pvcnn.py#L106)):*
<details><summary><i>Previously, ssheorey wrote…</i></summary><blockquote>

Add dtype and shape for return value. Is it (B, N, C)?
</blockquote></details>

Done.
___
*[ml3d/tf/models/pvcnn.py, line 112 at r2](https://reviewable.io/reviews/isl-org/open3d-ml/394#-MnMZ-6u4tghhG9rlMqj:-MnjlCWb2h8MPblsU-vH:bss4a2b) ([raw file](https://github.com/isl-org/open3d-ml/blob/835c152fc729f9aee6159ea2b23984a62a49f0d0/ml3d/tf/models/pvcnn.py#L112)):*
<details><summary><i>Previously, ssheorey wrote…</i></summary><blockquote>

> ```
> # coords = inputs[:, :3, :]
> ```
Is xyz dim1 or dim2? Update comment or docstring.
</blockquote></details>

Removed.
___
*[ml3d/torch/models/pvcnn.py, line 28 at r2](https://reviewable.io/reviews/isl-org/open3d-ml/394#-MnMRBHCBqV9T_VmdwUS:-MnjlGCl9mtZ2KhUoBZg:b-896fix) ([raw file](https://github.com/isl-org/open3d-ml/blob/835c152fc729f9aee6159ea2b23984a62a49f0d0/ml3d/torch/models/pvcnn.py#L28)):*
<details><summary><i>Previously, ssheorey wrote…</i></summary><blockquote>

Update docstrings to Google convention.
</blockquote></details>

Done.
___
*[ml3d/torch/models/pvcnn.py, line 137 at r2](https://reviewable.io/reviews/isl-org/open3d-ml/394#-MnMWChhCYdTAaWhZC6m:-MnjlHAH5oclh91LakmL:b-896fix) ([raw file](https://github.com/isl-org/open3d-ml/blob/835c152fc729f9aee6159ea2b23984a62a49f0d0/ml3d/torch/models/pvcnn.py#L137)):*
<details><summary><i>Previously, ssheorey wrote…</i></summary><blockquote>

Mention shape of output tensor. Is it `(B, N, C)`?
</blockquote></details>

Done.
___
*[ml3d/torch/models/pvcnn.py, line 143 at r2](https://reviewable.io/reviews/isl-org/open3d-ml/394#-MnMVmtnBhFvC7Q9V0lP:-MnjlHpyCQVGL-dv_QCM:ba3dfob) ([raw file](https://github.com/isl-org/open3d-ml/blob/835c152fc729f9aee6159ea2b23984a62a49f0d0/ml3d/torch/models/pvcnn.py#L143)):*
<details><summary><i>Previously, ssheorey wrote…</i></summary><blockquote>

> ```
> # coords = inputs[:, :3, :]
> ```
Are XYZ in dim 1 or dim 2? Update comment or docstring.
</blockquote></details>

Removed
___
*[ml3d/torch/models/pvcnn.py, line 188 at r2](https://reviewable.io/reviews/isl-org/open3d-ml/394#-MnMWhWO8l5z6lDK0QvK:-MnjlIyo9H2xUiME01FT:b-896fix) ([raw file](https://github.com/isl-org/open3d-ml/blob/835c152fc729f9aee6159ea2b23984a62a49f0d0/ml3d/torch/models/pvcnn.py#L188)):*
<details><summary><i>Previously, ssheorey wrote…</i></summary><blockquote>

> ```
> feat = feat / 255.0
> ```
Needs explanation.
</blockquote></details>

Done.
___
*[ml3d/torch/models/pvcnn.py, line 201 at r2](https://reviewable.io/reviews/isl-org/open3d-ml/394#-MnMWmQi72jzwajgr5S7:-MnjlJlC6uWBiULG6A-F:b-896fix) ([raw file](https://github.com/isl-org/open3d-ml/blob/835c152fc729f9aee6159ea2b23984a62a49f0d0/ml3d/torch/models/pvcnn.py#L201)):*
<details><summary><i>Previously, ssheorey wrote…</i></summary><blockquote>

Update with Numpy RNG PR and use `self.rng`.
</blockquote></details>

Done.
___
*[ml3d/torch/models/pvcnn.py, line 273 at r2](https://reviewable.io/reviews/isl-org/open3d-ml/394#-MnMXJmG3JN6CntnF9rc:-MnjlNG1-mhiHiU-GQtJ:b-896fix) ([raw file](https://github.com/isl-org/open3d-ml/blob/835c152fc729f9aee6159ea2b23984a62a49f0d0/ml3d/torch/models/pvcnn.py#L273)):*
<details><summary><i>Previously, ssheorey wrote…</i></summary><blockquote>

Loss -> sem_seg_loss
</blockquote></details>

Done.

**Review by benjaminum — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/isl-org/open3d-ml/394)* status: 0 of 34 files reviewed, 11 unresolved discussions (waiting on @benjaminum, @sanskar107, and @ssheorey)

___
*[ml3d/tf/models/pvcnn.py, line 594 at r3](https://reviewable.io/reviews/isl-org/open3d-ml/394#-Mo-7-CnDDuToulqWQmO:-Mo-7-CnDDuToulqWQmP:bdwey1i) ([raw file](https://github.com/isl-org/open3d-ml/blob/ff900aaefe44f3fe06e1a19f53802eaf976c56d8/ml3d/tf/models/pvcnn.py#L594)):*
> ```Python
>     """
> 
>     def __init__(self, resolution, normalize=True, eps=0):
> ```

Is `eps=0` a good default?
___
*[ml3d/torch/models/pvcnn.py, line 629 at r3](https://reviewable.io/reviews/isl-org/open3d-ml/394#-Mo-6PiDEP6Wmnw2KGqv:-Mo-6PiDEP6Wmnw2KGqw:bdwey1i) ([raw file](https://github.com/isl-org/open3d-ml/blob/ff900aaefe44f3fe06e1a19f53802eaf976c56d8/ml3d/torch/models/pvcnn.py#L629)):*
> ```Python
>     """
> 
>     def __init__(self, resolution, normalize=True, eps=0):
> ```

Is `eps=0` a good default?

**Review by sanskar107 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/isl-org/open3d-ml/394)* status: 0 of 34 files reviewed, 11 unresolved discussions (waiting on @benjaminum and @ssheorey)

___
*[ml3d/tf/models/pvcnn.py, line 594 at r3](https://reviewable.io/reviews/isl-org/open3d-ml/394#-Mo-7-CnDDuToulqWQmO:-Mo3lVEN7m0PqvK-m1Po:bg4zz59) ([raw file](https://github.com/isl-org/open3d-ml/blob/ff900aaefe44f3fe06e1a19f53802eaf976c56d8/ml3d/tf/models/pvcnn.py#L594)):*
<details><summary><i>Previously, benjaminum (Benjamin Ummenhofer) wrote…</i></summary><blockquote>

Is `eps=0` a good default?
</blockquote></details>

Added a small epsilon
___
*[ml3d/torch/models/pvcnn.py, line 629 at r3](https://reviewable.io/reviews/isl-org/open3d-ml/394#-Mo-6PiDEP6Wmnw2KGqv:-Mo3lRuVAEZrhOjzGDFr:bg4zz59) ([raw file](https://github.com/isl-org/open3d-ml/blob/ff900aaefe44f3fe06e1a19f53802eaf976c56d8/ml3d/torch/models/pvcnn.py#L629)):*
<details><summary><i>Previously, benjaminum (Benjamin Ummenhofer) wrote…</i></summary><blockquote>

Is `eps=0` a good default?
</blockquote></details>

Added a small epsilon

**Review by benjaminum — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/isl-org/open3d-ml/394#-:-Mo587Wu16JDjjE1gXI-:bnfp4nl)* status: 0 of 34 files reviewed, 11 unresolved discussions (waiting on @benjaminum and @ssheorey)

**Review by ssheorey — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/isl-org/open3d-ml/394#-:-MoEb9DB19CCtMzx6DGk:bnfp4nl)* status: 0 of 34 files reviewed, 11 unresolved discussions (waiting on @benjaminum and @ssheorey)
